bigchaindb/db/backends/rethinkdb.py

```diff
@@ -0,0 +1,381 @@
+        Args:
+            transaction_id (str): The id of the transaction.
+            condition_id (int): The index of the condition in the respective transaction.
+
+        Returns:
+            The transaction that used the `txid` as an input else `None`
+        """
+
+        # TODO: use index!
+        return self.connection.run(
+                r.table('bigchain', read_mode=self.read_mode)
+                .concat_map(lambda doc: doc['block']['transactions'])
+                .filter(lambda transaction: transaction['transaction']['fulfillments'].contains(
+                    lambda fulfillment: fulfillment['input'] == {'txid': transaction_id, 'cid': condition_id})))
+
+    def get_owned_ids(self, owner):
+        """Retrieve a list of `txids` that can we used has inputs.
+
+        Args:
+            owner (str): base58 encoded public key.
+
+        Returns:
+            A cursor for the matching transactions.
+        """
+
+        # TODO: use index!
+        return self.connection.run(
+                r.table('bigchain', read_mode=self.read_mode)
+                .concat_map(lambda doc: doc['block']['transactions'])
+                .filter(lambda tx: tx['transaction']['conditions'].contains(
+                    lambda c: c['owners_after'].contains(owner))))
+
+    def get_votes_by_block_id(self, block_id):
+        """Get all the votes casted for a specific block.
+
+        Args:
+            block_id (str): the block id to use.
+
+        Returns:
+            A cursor for the matching votes.
+        """
+        return self.connection.run(
+                r.table('votes', read_mode=self.read_mode)
+                .between([block_id, r.minval], [block_id, r.maxval], index='block_and_voter'))
+
+    def get_votes_by_block_id_and_voter(self, block_id, node_pubkey):
+        """Get all the votes casted for a specific block by a specific voter.
+
+        Args:
+            block_id (str): the block id to use.
+            node_pubkey (str): base58 encoded public key
+
+        Returns:
+            A cursor for the matching votes.
+        """
+        return self.connection.run(
+                r.table('votes', read_mode=self.read_mode)
+                .get_all([block_id, node_pubkey], index='block_and_voter'))
+
+    def write_block(self, block, durability='soft'):
+        """Write a block to the bigchain table.
+
+        Args:
+            block (dict): the block to write.
+
+        Returns:
+            The database response.
+        """
+        return self.connection.run(
+                r.table('bigchain')
+                .insert(r.json(block), durability=durability))
+
+    def has_transaction(self, transaction_id):
+        """Check if a transaction exists in the bigchain table.
+
+        Args:
+            transaction_id (str): the id of the transaction to check.
+
+        Returns:
+            ``True`` if the transaction exists, ``False`` otherwise.
+        """
+        return bool(self.connection.run(
```

Comment:
This is correct but I just want to make a note because `bool` always makes me nervous. There are a lot of values that evaluate to true and this can lead to undefined behavior, especially when the input of `bool` comes from an external call